tests/integration/backward_compatible/ssl_tests/hostname_verification/ssl_hostname_verification_test.py

```diff
@@ -0,0 +1,128 @@
+        self.start_member_with("tls-host-loopback-san.p12")
+
+        self.start_client_with("tls-host-loopback-san.pem", "127.0.0.1:5701")
+        self.start_client_with("tls-host-loopback-san.pem", "localhost:5701")
+
+    def test_hostname_verification_with_loopback_dns_san(self):
+        # SAN entry is present, but only with `dns:localhost`
+        self.start_member_with("tls-host-loopback-san-dns.p12")
+        self.start_client_with("tls-host-loopback-san-dns.pem", "localhost:5701")
+
+        with self.assertRaisesRegex(IllegalStateError, "Unable to connect to any cluster"):
+            self.start_client_with("tls-host-loopback-san-dns.pem", "127.0.0.1:5701")
+
+    def test_hostname_verification_with_different_san(self):
+        # There is a valid entry, but it does not match with the address of the member.
+        self.start_member_with("tls-host-not-our-san.p12")
+
+        with self.assertRaisesRegex(IllegalStateError, "Unable to connect to any cluster"):
+            self.start_client_with("tls-host-not-our-san.pem", "localhost:5701")
+
+        with self.assertRaisesRegex(IllegalStateError, "Unable to connect to any cluster"):
+            self.start_client_with("tls-host-loopback-san-dns.pem", "127.0.0.1:5701")
+
+    def test_hostname_verification_with_loopback_cn(self):
+        # No entry in SAN but an entry in CN which checked as a fallback
+        # when no entry in SAN is present.
+        self.start_member_with("tls-host-loopback-cn.p12")
+
+        self.start_client_with("tls-host-loopback-cn.pem", "localhost:5701")
+
+        # See https://stackoverflow.com/a/8444863/12394291. IP addresses in CN
+        # are not supported. So, we don't have a test for it.
+        with self.assertRaisesRegex(IllegalStateError, "Unable to connect to any cluster"):
+            self.start_client_with("tls-host-loopback-cn.pem", "127.0.0.1:5701")
+
+    def test_hostname_verification_with_no_entry(self):
+        # No entry either in the SAN or CN. No way to verify hostname.
+        self.start_member_with("tls-host-no-entry.p12")
+
+        with self.assertRaisesRegex(IllegalStateError, "Unable to connect to any cluster"):
+            self.start_client_with("tls-host-no-entry.pem", "localhost:5701")
+
+        with self.assertRaisesRegex(IllegalStateError, "Unable to connect to any cluster"):
+            self.start_client_with("tls-host-no-entry.pem", "127.0.0.1:5701")
+
+    def test_hostname_verification_disabled(self):
+        # When hostname verification is disabled, the scenarious that
+        # would fail in `test_hostname_verification_with_no_entry` will
+        # no longer fail, showing that it is working as expected.
+        self.start_member_with("tls-host-no-entry.p12")
+
+        self.start_client_with("tls-host-no-entry.pem", "localhost:5701", check_hostname=False)
+        self.start_client_with("tls-host-no-entry.pem", "127.0.0.1:5701", check_hostname=False)
```

Comment:
Do you think we can use table driven tests to reduce the test code? https://docs.python.org/3/library/unittest.html#unittest.TestCase.subTest

Reply:
I took a look at it and it looks cool but, I decided not to change the tests to that because I think it decreases the readability and our ability to debug the tests. Besides, I have some comments on the tests that explain the test a little bit which could be helpful for future readers.